### PR TITLE
Add unique key

### PR DESCRIPTION
**Pull request by alexalouit**
unique key for mysql
unique key for sqlite

### CODE REVIEW

**Comment by alexalouit** [comment]
we need to add unique key for pgsql

File: daos/sqlite/Database.php [hunk 1]
```diff
@@ -235,9 +235,10 @@ public function __construct() {
                     ]);
                 }
                 if (strnatcmp($version, '12') < 0) {
-                    \F3::get('db')->exec('
-                        CREATE UNIQUE INDEX uid ON new_items(uid, source);
-                    ');
+                    \F3::get('db')->exec([
+                        'CREATE UNIQUE INDEX uid ON new_items(uid, source);',
+                        'INSERT INTO version (version) VALUES (12)'
```
Comment (alexalouit) [inline daos/sqlite/Database.php]: I'm not sure for the length of index, pgsql support it?

File: daos/sqlite/Database.php [hunk 2]
```diff
@@ -234,6 +234,12 @@ public function __construct() {
                         'INSERT INTO version (version) VALUES (11)'
                     ]);
                 }
+                if (strnatcmp($version, '12') < 0) {
+                    \F3::get('db')->exec([
+                        'CREATE UNIQUE INDEX uid ON new_items(uid, source);',
```
Comment (jtojnar) [inline daos/sqlite/Database.php]: The index should be created on `items`, there should be no table `new_items` at this point.

File: daos/sqlite/Database.php [hunk 3]
```diff
@@ -234,6 +234,12 @@ public function __construct() {
...
+                        'INSERT INTO version (version) VALUES (12)'
```
Comment (jtojnar) [inline daos/sqlite/Database.php]: Could you use version 13 so that it is more clear that the database is current.

File: daos/pgsql/Database.php [hunk 4]
```diff
@@ -227,6 +227,12 @@ public function __construct() {
                         'INSERT INTO version (version) VALUES (12)'
                     ]);
                 }
+                if (strnatcmp($version, '13') < 0) {
+                    \F3::get('db')->exec([
+                        'ALTER TABLE ' . \F3::get('db_prefix') . 'items UNIQUE (uid (150), source);',
+                        'INSERT INTO ' . \F3::get('db_prefix') . 'version (version) VALUES (13)'
```
Comment (jtojnar) [inline daos/pgsql/Database.php]: There is no need to use `db_prefix` in Postgres. It is a measure for cheap shared hostings that only provide a single MySQL database.

File: daos/pgsql/Database.php [hunk 5]
```diff
@@ -227,6 +227,12 @@ public function __construct() {
                         'INSERT INTO version (version) VALUES (12)'
                     ]);
                 }
+                if (strnatcmp($version, '13') < 0) {
+                    \F3::get('db')->exec([
+                        'ALTER TABLE ' . \F3::get('db_prefix') . 'items UNIQUE (uid (150), source);',
```
Comment (jtojnar) [inline daos/pgsql/Database.php]: PostgreSQL is smart enough that no length should be necessary. Additionally, I think the syntax is `ALTER TABLE items ADD UNIQUE(uid, source)`, see the definitions for [`ALTER TABLE`](https://www.postgresql.org/docs/current/static/sql-altertable.html) and [`table_constraint`](https://www.postgresql.org/docs/current/static/sql-createtable.html).

File: daos/mysql/Database.php [hunk 6]
```diff
@@ -243,6 +243,12 @@ public function __construct() {
                         'INSERT INTO ' . \F3::get('db_prefix') . 'version (version) VALUES (12)'
                     ]);
                 }
+                if (strnatcmp($version, '13') < 0) {
+                    \F3::get('db')->exec([
+                        'ALTER IGNORE TABLE ' . \F3::get('db_prefix') . 'items ADD UNIQUE INDEX (uid (150), source);',
```
Comment (jtojnar) [inline daos/mysql/Database.php]: It looks like `ALTER IGNORE TABLE` syntax is [not supported](https://dev.mysql.com/doc/refman/5.6/en/alter-table.html#idm139885224435136) any more in MySQL 5.7. Manual deduplication will be needed.

Comment (jtojnar) [inline daos/mysql/Database.php]: Also I am not sure about the 150 byte length. Would not 255 work as well? That should fit 1024 bytes even with four bytes per character.

File: daos/sqlite/Database.php (hunk 2 above)
Comment (alexalouit) [reply daos/sqlite/Database.php]: 7040883bfbd248dae6c5d266912379c31cfcacd5

File: daos/sqlite/Database.php (hunk 3 above)
Comment (alexalouit) [reply daos/sqlite/Database.php]: fix in 7040883bfbd248dae6c5d266912379c31cfcacd5

File: daos/pgsql/Database.php (hunk 4 above)
Comment (alexalouit) [reply daos/pgsql/Database.php]: fix in 7040883bfbd248dae6c5d266912379c31cfcacd5

File: daos/pgsql/Database.php (hunk 5 above)
Comment (alexalouit) [reply daos/pgsql/Database.php]: fix in 7040883bfbd248dae6c5d266912379c31cfcacd5

File: daos/mysql/Database.php (hunk 6 above)
Comment (alexalouit) [reply daos/mysql/Database.php]: fix in 7040883bfbd248dae6c5d266912379c31cfcacd5

Comment (alexalouit) [reply daos/mysql/Database.php]: On MariaDB 5.5, I got: ```Specified key was too long; max key length is 767 bytes```
It only works with less than ~190.

see https://stackoverflow.com/questions/1814532/1071-specified-key-was-too-long-max-key-length-is-767-bytes/1814594#1814594

Comment (jtojnar) [reply daos/mysql/Database.php]: Oh, InnoDB. 191 should be used then.

Comment (alexalouit) [reply daos/mysql/Database.php]: fix in 2cc1cb8

**Comment by jtojnar** [comment]
Rebased this onto master and resolved conflicts.

But yeah, this still needs to handle past duplicates otherwise it will fail with `PDOStatement: UNIQUE constraint failed: items.uid, items.source`

**Pull request closed without merging**